Add ML library versions to pickle file for compatibility

rhoboro opened the pull request:
⚠️ This PR will be directed to the pycaret/pycaret repository. ⚠️ 

## Related Issue or bug

Info about Issue or bug

Closes #[issue number that will be closed through this PR]

#### Describe the changes you've made

PyCaret 2.x uses sklearn 0.23.2, and PyCaret 3.x will always use the latest version of sklearn.
Since the versions of the various ML libraries are no longer fixed, this PR will embed the ML library versions (including sklearn) in the pickle file.
This PR was created with reference to [sklearn/base.py](https://github.com/scikit-learn/scikit-learn/blob/8175cd60e3df606905aedf2dd3ba0832e50cd963/sklearn/base.py#L326) :)

Here is an example.

```sh
# refs. https://pycaret.gitbook.io/docs/get-started/quickstart#classification
$ python3
>>> from pycaret.datasets import get_data
>>> data = get_data('diabetes')
>>> from pycaret.classification import *
>>> s = setup(data, target = 'Class variable')
>>> best = compare_models()
>>> save_model(best, 'my_best_pipeline')
>>> 
# This works fine, of course. 
>>> loaded_model = load_model('my_best_pipeline')
Transformation Pipeline and Model Successfully Saved
(Pipeline(memory=Memory(location=/var/folders/sj/n8jwy9zx4vq74qw53_3m7lxr0000gn/T/joblib),
         steps=[('numerical_imputer',
                 TransformerWrapper(exclude=None, include=['Number of times pregnant', 'Plasma glucose concentration a 2 hours in an oral glucose tolerance test', 'Diastolic blood pressure (mm Hg)', 'Triceps skin fold thickness (mm)', '2-Hour serum insulin (mu U/ml)', 'Body mass inde...
                 TransformerWrapper(exclude=None, include=[], transformer=SimpleImputer(add_indicator=False, copy=True, fill_value='constant', missing_values=nan, strategy='constant', verbose='deprecated'))),
                ['trained_model',
                 LinearDiscriminantAnalysis(covariance_estimator=None,
                                            n_components=None, priors=None,
                                            shrinkage=None, solver='svd',
                                            store_covariance=False,
                                            tol=0.0001)]],
         verbose=False), 'my_best_pipeline.pkl')

# loaded_model has its own version information
>>> loaded_model._pycaret_versions
{'pycaret': '3.0.0.rc3', 'sklearn': '1.1.1', 'imblearn': '0.9.1', 'pyod': '1.0.2', 'category-encoders': '2.5.0', 'lightgbm': '3.3.2'}
```

This will increase the availability of the pycaret models.
In an environment with the latest pycaret installed, the following snippet shows the version of the pkl file when it was created.
This allows us to recreate an environment in which the model can be executed.

```sh
# In a virtual environment with a fake updated version of pycaret installed
$ python3 peeker_sample.py my_best_pipeline
Transformation Pipeline and Model Successfully Loaded
Version mismatch:
current: {'pycaret': '3.0.0.rc4', 'sklearn': '1.1.1', 'imblearn': '0.9.1', 'pyod': '1.0.3', 'category-encoders': '2.5.0', 'lightgbm': '3.3.2'}
pickle: {'pycaret': '3.0.0.rc3', 'sklearn': '1.1.1', 'imblearn': '0.9.1', 'pyod': '1.0.2', 'category-encoders': '2.5.0', 'lightgbm': '3.3.2'}
```

```peeker_sample.py
# peeker_sample.py
import sys
import warnings

from pycaret.classification import *


def check(target):
    with warnings.catch_warnings(record=True) as w:
        warnings.simplefilter("always")
        loaded_model = load_model(target)
        for m in w:
            print(m.message)


def main():
    check(sys.argv[1])


if __name__ == "__main__":
    main()
```

#### Note

If model_only=True is passed to save_model(), this change does nothing.
In this case, it is up to the library implementation to provide version information.

ex. sklearn model stores that version.

```
>>> save_model(best, 'my_best_pipeline_model_only', model_only=True)
Transformation Pipeline and Model Successfully Saved
(RidgeClassifier(alpha=1.0, class_weight=None, copy_X=True, fit_intercept=True,
                max_iter=None, normalize='deprecated', positive=False,
                random_state=4741, solver='auto', tol=0.001), 'my_best_pipeline_model_only.pkl')
>>> loaded_model = load_model('my_best_pipeline_model_only')
Transformation Pipeline and Model Successfully Loaded
>>> type(loaded_model)
<class 'sklearn.linear_model._ridge.RidgeClassifier'>
>>> loaded_model.__getstate__()['_sklearn_version']
'1.1.1'
```


## Type of change

Please delete options that are not relevant.
<!--
Example how to mark a checkbox :-
- [x] My code follows the code style of this project.
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, local variables)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Manual testing was done with quick start (classification, regression, clustering, anomaly-detection).

## Checklist:

<!--
Example how to mark a checkbox :-
- [x] My code follows the code style of this project.
-->
- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

## Screenshots

 Original           | Updated
 :--------------------: |:--------------------:
 **original screenshot**  | <b>updated screenshot </b> |

rhoboro committed:
Add ML library versions into the pickle file

kennysong commented on pycaret/internal/pipeline.py:
```diff
@@ -79,6 +80,42 @@ def __init__(self, steps, *, memory=None, verbose=False):
         self._fit_vars = set()
```
Just want to double check that `class Pipeline(imblearn.pipeline.Pipeline)` is the class that's used for all PyCaret classification and regression models created by `compare_models()` or `create_model()`?

rhoboro replied:
That is my understanding.
Each `setup()` functions initialize ``pycaret.internal.pipeline.Pipeline` in following rows.
- regression
    - regression/oop.py: 678
- classification
    - classification/oop.py: 709
- anomaly/clustering
    - unsupervised_experiment.py: 694

kennysong commented on pycaret/internal/pipeline.py:
```diff
@@ -79,6 +80,42 @@ def __init__(self, steps, *, memory=None, verbose=False):
+
+    def __setstate__(self, state):
+        pickle_versions = state.get("_pycaret_version", {})
+        if pickle_versions["pycaret"] != __version__:
```
This will fail if pickle_versions is `{}` right?

```suggestion
        if pickle_versions.get("pycaret") != __version__:
```

kennysong commented on pycaret/internal/pipeline.py:
```diff
@@ -79,6 +80,42 @@ def __init__(self, steps, *, memory=None, verbose=False):
+        if pickle_versions["pycaret"] != __version__:
+            import warnings
+
+            warnings.warn("Version mismatch: {}".format(pickle_versions))
```
For clarity, I suggest writing a more detailed warning here with the current version and pickle version.

rhoboro replied:
fixed
https://github.com/rhoboro/pycaret/pull/1/commits/ffab617fe3a65c0288f57ce4fc8d4c34e82203fc

kennysong commented on pycaret/internal/pipeline.py:
```diff
@@ -79,6 +80,42 @@ def __init__(self, steps, *, memory=None, verbose=False):
         self._feature_names_in = None
 
+    def __getstate__(self):
+        try:
```
I don't understand this if/else block.

When would `super().__getstate__()` raise an `AttributeError`? 

Also, by default, this seems to only save the state from the parent `imblearn.pipeline.Pipeline` and discards the state from the actual `Pipeline`?

rhoboro replied:
The `__getstate__()` is an optional special method, so super class does not necessarily have it. ( However, the current superclass, imblearn.pipeline.Pipeline, has it. )

The `self` in superclass's methods has attributes defined in its subclasses.
Thus, when the superclass executes `state = self.__dict__.copy()`, all attributes are saved.
However, since there is no guarantee that `state = self.__dict__.copy()` will be executed, I  commited some changes.

kennysong commented on pycaret/internal/pipeline.py:
```diff
@@ -79,6 +80,42 @@ def __init__(self, steps, *, memory=None, verbose=False):
+            import warnings
+
+            warnings.warn("Version mismatch: {}".format(pickle_versions))
+        try:
```
Is this the standard way to set state from a pickled object?

rhoboro replied:
I do not know of a standard way, but sklearn implements this method.

https://github.com/scikit-learn/scikit-learn/blob/6e7922e48a04f471b11e874fac67657941970794/sklearn/base.py#L326-L344

kennysong reviewed (commented):
Generally looks good, but I left some questions. Thanks Suyama-san!

AseiSugiyama commented:
TODO

- [ ] Check the behavior without PyCarret (especially 3.0.rc0)
- ~~Check the behavior with PyCarret 2.x~~

rhoboro commented:
@AseiSugiyama 

> Check the behavior without PyCarret (especially 3.0.rc0)
> Check the behavior with PyCarret 2.x

It is recommended to always load with the latest pycaret.
If the versions match, the load will succeed; if not, a warning and the version used will be displayed.

rhoboro commented:
@kennysong @AseiSugiyama 
Thank you for the review.

I have rebased to the latest master branch and updated the PR description. Please check it.
If everything looks OK, I will create a PR in the pycaret/pycaret repository.
(I didn't have time to check after rebasing, so I will it first.)

kennysong commented:
I've manually tested the latest version and it's working! The description also looks good. You can link to this issue in the "Related Issue or Bug" section: https://github.com/pycaret/pycaret/issues/2016

Looks good to submit the PR!
